### PR TITLE
Drop support for Doctrine cache adapter for metadata

### PR DESCRIPTION
**Pull request by GromNaN**
Remove BC layer introduced in #690 to keep support for Doctrine cache adapters. PSR-6 is now required.

### CODE REVIEW

**Comment by GromNaN** [comment]
@Ahummeling, you introduced this compatibility layer, I'd be glad to have your review if you can find the time.

**Comment by franmomu** [comment]
There are also these class definitions involved:

https://github.com/doctrine/DoctrineMongoDBBundle/blob/ddf39bb974080d82325eff633dba7ce8c874a5ad/Resources/config/mongodb.xml#L22-L30

I don't know (I haven't properly checked) if changing them would also allow us to remove:

https://github.com/doctrine/DoctrineMongoDBBundle/blob/ddf39bb974080d82325eff633dba7ce8c874a5ad/DependencyInjection/DoctrineMongoDBExtension.php#L580-L650

since apparently these classes are used by [AbstractDoctrineExtension](https://github.com/symfony/symfony/blob/e0df5c3d9c0281cdfe3885281145e2065b859d32/src/Symfony/Bridge/Doctrine/DependencyInjection/AbstractDoctrineExtension.php#L340-L406).

**Comment by Ahummeling** [comment]
> @Ahummeling, you introduced this compatibility layer, I'd be glad to have your review if you can find the time.

It's been a while since I've touched this code, I might be able to find some time over the weekend. It was quite a complex task, so would like to take the time for it to properly review this.

**Review by alcaeus — APPROVED** [review]
LGTM. Do we have any other remnants of doctrine/cache to clean up?

File: DependencyInjection/DoctrineMongoDBExtension.php [hunk 1]
```diff
@@ -582,55 +580,57 @@ protected function loadCacheDriver(string $cacheName, string $objectManagerName,
...
                 $memcachedInstance->addMethodCall('addServer', [
                     $memcachedHost,
                     $memcachedPort,
                 ]);
                 $container->setDefinition($this->getObjectManagerElementName(sprintf('%s_memcached_instance', $objectManagerName)), $memcachedInstance);
 
-                $cacheDef = new Definition(MemcachedAdapter::class, [new Reference($this->getObjectManagerElementName(sprintf('%s_memcached_instance', $objectManagerName)))]);
+                $cacheDef = new Definition($memcachedClass, [
+                    new Reference($this->getObjectManagerElementName(sprintf('%s_memcached_instance', $objectManagerName))),
+                    $cacheDriver['namespace'] ?? '',
```
Comment (GromNaN) [inline DependencyInjection/DoctrineMongoDBExtension.php]: Namespace config can be backported and tested.

File: DependencyInjection/DoctrineMongoDBExtension.php [hunk 2]
```diff
@@ -582,55 +580,57 @@ protected function loadCacheDriver(string $cacheName, string $objectManagerName,
...
                 $redisInstance->addMethodCall('connect', [
                     $redisHost,
                     $redisPort,
                 ]);
                 $container->setDefinition($this->getObjectManagerElementName(sprintf('%s_redis_instance', $objectManagerName)), $redisInstance);
 
-                $cacheDef = new Definition(RedisAdapter::class, [new Reference($this->getObjectManagerElementName(sprintf('%s_redis_instance', $objectManagerName)))]);
+                $cacheDef = new Definition($redisClass, [
+                    new Reference($this->getObjectManagerElementName(sprintf('%s_redis_instance', $objectManagerName))),
+                    $cacheDriver['namespace'] ?? '',
```
Comment (GromNaN) [inline DependencyInjection/DoctrineMongoDBExtension.php]: Namespace config can be backported and tested.

File: DependencyInjection/DoctrineMongoDBExtension.php [hunk 3]
```diff
@@ -582,55 +580,57 @@ protected function loadCacheDriver(string $cacheName, string $objectManagerName,
...
                 break;
 
             case 'redis':
-                if (! empty($cacheDriver['class']) && $cacheDriver['class'] !== RedisCache::class) {
-                    return parent::loadCacheDriver($cacheName, $objectManagerName, $cacheDriver, $container);
-                }
-
-                $redisInstanceClass = ! empty($cacheDriver['instance_class']) ? $cacheDriver['instance_class'] : '%' . $this->getObjectManagerElementName('cache.redis_instance.class') . '%';
-                $redisHost          = ! empty($cacheDriver['host']) ? $cacheDriver['host'] : '%' . $this->getObjectManagerElementName('cache.redis_host') . '%';
-                $redisPort          = ! empty($cacheDriver['port']) ? $cacheDriver['port'] : '%' . $this->getObjectManagerElementName('cache.redis_port') . '%';
```
Comment (GromNaN) [inline DependencyInjection/DoctrineMongoDBExtension.php:609]: This parameters are not defined, we should deprecate them.

File: DependencyInjection/DoctrineMongoDBExtension.php [hunk 4]
```diff
@@ -582,55 +580,57 @@ protected function loadCacheDriver(string $cacheName, string $objectManagerName,
                 return $cacheDriverServiceId;
 
             case 'memcached':
-                if (! empty($cacheDriver['class']) && $cacheDriver['class'] !== MemcacheCache::class) {
-                    return parent::loadCacheDriver($cacheName, $objectManagerName, $cacheDriver, $container);
-                }
-
-                $memcachedInstanceClass = ! empty($cacheDriver['instance_class']) ? $cacheDriver['instance_class'] : '%' . $this->getObjectManagerElementName('cache.memcached_instance.class') . '%';
-                $memcachedHost          = ! empty($cacheDriver['host']) ? $cacheDriver['host'] : '%' . $this->getObjectManagerElementName('cache.memcached_host') . '%';
-                $memcachedPort          = ! empty($cacheDriver['port']) ? $cacheDriver['port'] : '%' . $this->getObjectManagerElementName('cache.memcached_port') . '%';
```
Comment (GromNaN) [inline DependencyInjection/DoctrineMongoDBExtension.php:590]: This parameters needs to be deprecated.

File: DependencyInjection/DoctrineMongoDBExtension.php [hunk 5]
```diff
@@ -582,55 +580,57 @@ protected function loadCacheDriver(string $cacheName, string $objectManagerName,
...
+                $cacheDef = new Definition($redisClass, [
+                    new Reference($this->getObjectManagerElementName(sprintf('%s_redis_instance', $objectManagerName))),
+                    $cacheDriver['namespace'] ?? '',
+                ]);
 
                 break;
 
             case 'apcu':
-                $cacheDef = new Definition(ApcuAdapter::class);
+                $cacheDef = new Definition($cacheDriver['class'] ?? ApcuAdapter::class, [
```
Comment (GromNaN) [inline DependencyInjection/DoctrineMongoDBExtension.php]: This a new feature at no cost: customize the class and set the namespace.

File: Resources/config/mongodb.xml [hunk 6]
```diff
@@ -49,7 +39,6 @@
 
     <services>
         <!-- defaults -->
-        <service id="doctrine_mongodb.odm.cache" alias="doctrine_mongodb.odm.cache.array" />
```
Comment (GromNaN) [inline Resources/config/mongodb.xml]: Should be deprecated in 4.7. This is not used.

File: DependencyInjection/DoctrineMongoDBExtension.php (hunk 5 above)
Comment (GromNaN) [reply DependencyInjection/DoctrineMongoDBExtension.php]: Removed as useless.

**Comment by GromNaN** [comment]
@alcaeus @franmomu I updated the extension and service definitions to not rely on the abstract extension from doctrine-bridge.

**Review by alcaeus — APPROVED** [review]

**Pull request merged**